### PR TITLE
[Access] Support fetching system tx result from cache

### PR DESCRIPTION
**Pull request by illia-malachyn**
Cache system tx result. Add test for it.

Issue: https://github.com/onflow/flow-go/issues/5757

### CODE REVIEW

**Review by peterargue — COMMENTED** [review]
thanks for the PR @illia-malachyn!

This issue is actually about adding support for querying the transaction from a the local database when indexing is available. You can probably make use of this method to handle the logic of deciding how to get the data:
https://github.com/onflow/flow-go/blob/8fac7a40c49d4f020fe1a5adbe56980128d9238a/engine/access/rpc/backend/backend_transactions.go#L667-L702

**Comment by illia-malachyn** [comment]
> thanks for the PR @illia-malachyn!
> 
> This issue is actually about adding support for querying the transaction from the local database when indexing is available. You can probably make use of this method to handle the logic of deciding how to get the data:
> 
> https://github.com/onflow/flow-go/blob/8fac7a40c49d4f020fe1a5adbe56980128d9238a/engine/access/rpc/backend/backend_transactions.go#L667-L702

Hey. Thanks! 
I've rewritten that code

File: engine/access/rpc/backend/backend_transactions.go [hunk 1]
```diff
@@ -801,6 +782,7 @@ func (b *backendTransactions) getTransactionResultFromExecutionNode(
 		return nil, rpc.ConvertStorageError(err)
 	}
 
+	// events, err := convert.MessagesToEventsWithEncodingConversion(systemTxResult.GetEvents(), resp.GetEventEncodingVersion(), requiredEventEncodingVersion)
```
Comment (peterargue) [inline engine/access/rpc/backend/backend_transactions.go:785]:
```suggestion
```

File: engine/access/rpc/backend/backend_transactions.go [hunk 2]
```diff
@@ -609,42 +609,23 @@ func (b *backendTransactions) GetSystemTransactionResult(ctx context.Context, bl
...
 	}
 
-	req := &execproto.GetTransactionsByBlockIDRequest{
-		BlockId: blockID[:],
-	}
-	execNodes, err := executionNodesForBlockID(ctx, blockID, b.executionReceipts, b.state, b.log)
-	if err != nil {
-		if IsInsufficientExecutionReceipts(err) {
-			return nil, status.Errorf(codes.NotFound, err.Error())
+	if b.txResultCache != nil {
```
Comment (peterargue) [inline engine/access/rpc/backend/backend_transactions.go:612]: I think caching this is generally a good idea, however this cache is specifically used for historic tx results and is generally not used. my preference is to not cache here, and return the result from `lookupTransactionResult`

File: engine/access/rpc/backend/backend_transactions_test.go [hunk 3]
```diff
@@ -893,11 +898,100 @@ func (suite *Suite) TestGetSystemTransactionResult_HappyPath() {
...
+	})
+}
+
+// TestGetSystemTransactionResultFromCacheWhenFound tests that GetSystemTransactionResult call
+// returns system transaction result for required block id from the cache.
+//
+// How it is tested: In case of cache hit, GetSystemTransactionResult doesn't
+// reach out to the execution nodes and returns early, so we check how
+// many times we do reach execution nodes
+func (suite *Suite) TestGetSystemTransactionResultFromCacheWhenFound() {
```
Comment (peterargue) [inline engine/access/rpc/backend/backend_transactions_test.go:922]: this test should cover looking up the result from storage. it's basically the same test we have for tx results, just using the system tx.

**Comment by illia-malachyn** [comment]
Unit tests (engine) fail for no reason? Locally everything is fine. However, Uliana noticed that `TestExecuteBlockInOrder` (from engine folder) fails sometimes, maybe it is a flaky one.

**Pull request merged**